arroyo/processing/strategies/run_task_with_multiprocessing.py

```diff
@@ -604,12 +646,54 @@ def __check_for_results_impl(self, timeout: Optional[float] = None) -> None:
                         result.next_index_to_process,
                     ),
                 ),
+                input_block_too_small,
+                True,
             )
             return
 
+        old_input_block = input_batch.block
+
+        if (
+            input_block_too_small
+            and self.__resize_input_blocks
+            and (
+                self.__max_input_block_size is None
+                or self.__max_input_block_size > old_input_block.size * 2
+            )
+        ):
+            self.__metrics.increment(
+                "arroyo.strategies.run_task_with_multiprocessing.batch.input.resize"
+            )
+            new_input_block = self.__shared_memory_manager.SharedMemory(
```

Comment:
Knowing that memory settings for Kubernetes are static (we cannot resize the memory assigned to a pod) and that you cannot exceed the memory allocated (OOMKill), do we have a reason to ever resize our input/output blocks instead of just taking all the available memory at the start ? It seems it would be much easier to just statically create memory blocks that used all the available memory of the pod and not try to change anything.

Reply:
> do we have a reason to ever resize our input/output blocks instead of just taking all the available memory at the start

there's still a ton of stuff outside of arroyo that takes memory, but even if not I don't think it's a good idea to consume significantly more memory than needed. When we deploy a new service we configure k8s limits based on projected memory usage, and later optionally adjust based on average memory usage. If arroyo defaults to consuming all pod memory, we lose insight into how much memory we actually need and how much cost we could save.

Reply:
> but even if not I don't think it's a good idea to consume significantly more memory than needed

Why?  That memory is already allocated and not usable by others anyway. 
If you are concerned of not having visibility on the actual usage, why not having specific metrics for that? It seems an easier, safer system with fewer moving parts and fewer failure modes.

Reply:
The goal is to eliminate tuning parameters the user has to tweak to get optimal consumer performance. If we manage to do that, then we can start thinking of removing some of those options (as they are not required and have optimal defaults) and therefore moving parts in that sense.

An engineer of the product team should not have to think about how much memory their pod is going to consume and tune Arroyo parameters based off of it.

I don't think this is possible at all with a static approach, because it requires the author of the consumer to think about how much memory their pod has (unclear, gets adjusted by ops), how much their regular code consumes per-process (entirely unclear, especially in a shared codebase like sentry where tons of random stuff gets imported at every CLI invocation) and then think about how much of the remainder can be allocated to input/output blocks.

If you are suggesting a static approach that is also zero-config, I don't know how that would work. Does it mean that arroyo determines free memory and allocates it evenly divided for input/output blocks? And is it evenly, or do input blocks get more than output blocks? And what does it do on a dev machine where there's no k8s request/limit per-consumer?

Reply:
> If you are concerned of not having visibility on the actual usage, why not having specific metrics for that?

I think this is possible but it feels like Java/node heap tuning parameters and I would like to avoid that sort of experience as well.